integrations/go/workload_executor_test.go

```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"testing"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
+	"go.mongodb.org/mongo-driver/mongo/integration/unified"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
+// hasLoop returns whether or not testCase tc contains a loop operation
+func hasLoop(tc *unified.TestCase) bool {
+	if tc == nil {
+		return false
+	}
+	for _, op := range tc.Operations {
+		if op.Name == "loop" {
+			return true
+		}
+	}
+	return false
+}
+
+func TestAtlasPlannedMaintenance(t *testing.T) {
+	connstring := os.Args[1]
+	workloadSpec := []byte(os.Args[2])
+
+	setupOpts := mtest.NewSetupOptions().SetURI(connstring)
+	if err := mtest.Setup(setupOpts); err != nil {
+		panic(err)
+	}
+	defer func() {
+		if err := mtest.Teardown(); err != nil {
+			panic(err)
+		}
+	}()
+
+	// killAllSessions will return an auth error if it's run
+	fileReqs, runners := unified.ParseTestFile(t, workloadSpec, unified.NewOptions().SetRunKillAllSessions(false))
+
+	mtOpts := mtest.NewOptions().
+		RunOn(fileReqs...).
+		CreateClient(false)
+	mt := mtest.New(t, mtOpts)
+	defer mt.Close()
+
+	// a testfile can parse to multiple runners, though we only expect to get one per file
+	for _, runner := range runners {
+		mtOpts := mtest.NewOptions().
+			RunOn(runner.RunOnRequirements...).
+			CreateClient(false)
+
+		mt.RunOpts(runner.Description, mtOpts, func(mt *mtest.T) {
+			// the workload executor should be able to run non-looping tests and EndLoop() will panic
+			// if the test has already finished
+			if hasLoop(runner) {
+				// Waits for the termination signal from astrolabe and terminates the loop operation
+				go func() {
+					c := make(chan os.Signal, 1)
+					signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+
+					<-c
+					runner.EndLoop()
+				}()
+			}
+
+			testErr := runner.Run(mt)
+
+			em := runner.GetEntities()
+
+			// structs for marshaling json results
+			results := struct {
```

Comment:
All of this code looks fine, but from an organizational standpoint, I think it could be reordered for readability. I'm thinking something like:

```
var allEvents struct {
  // fields
}

// Same logic to set allEvents failures, errors, and events fields using entity map and handle testErr
// Marshal allEvents to JSON and write to file

var results struct { ... }
// use allEvents to derive fields for results
// Marshal results to JSON and write to file
```

This way, we finish working with `allEvents` before moving on to `results` rather than bouncing between them in the code.

Reply:
done